### ISSUE TITLE
Fix pyflakes for trial

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7005]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#7005|
|Type|enhancement|
|Created|2014-02-27 14:22:55Z|
|Branch|https://github.com/twisted/twisted/tree/pyflakes-for-trial-7005|

Attachments:

* [7005-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/040/040cf8079ba56b68c2c75728b22904eebe31ee87/437bd685676f5dabbd268e4f362030eab4408237.diff) (10462 bytes) - added by adiroiban on 2014-02-27 14:37:38Z - 
* [7005-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/040/040cf8079ba56b68c2c75728b22904eebe31ee87/72d2485417b82aa6d0249103f4bb9ecf21de2cef.diff) (15096 bytes) - added by adiroiban on 2014-04-03 13:19:07Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7005 7005
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__ 
branch__branches_pyflakes_for_trial_7005 branches/pyflakes-for-trial-7005
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1393510975000000 1393510975000000
changetime__1423234419295894 1423234419295894
version__None None
owner__adiroiban adiroiban
cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban commented|
|-|-|

Only minor cleaup of unused imports and try to explain pyflakes that an import is used as a side effect.

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|Alex commented|
|-|-|

In my opinion a much better approach to silences the linter on unused imports would be to make use of `__all__`, which `flake8` respects.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Look again - these aren't top-level imports.

This seems like a pyflakes bug to me.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

I have attached a new patch which used namedAny and updates __all__ member.

Please take another look.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_6"></a>|@wallrj commented|
|-|-|

(In [42639]) Branching to 'pyflakes-for-trial-7005'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_7"></a>|@wallrj commented|
|-|-|

(In [42640]) Apply 7005-2.diff from adiroban. Refs [#7005](https://github.com/twisted/twisted/issues/7005)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_8"></a>|@wallrj commented|
|-|-|

Thanks Adi,

Notes:
 * Forced a build: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/pyflakes-for-trial-7005
   * Python3 build fails. On the face of it, it looks unrelated, but it's confusing that the trunk build passes. https://buildbot.twistedmatrix.com/builders/python-3.3-tests/builds/2229/steps/shell/logs/stdio
 * Tests pass locally

I'll try and finish off the review later.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_9"></a>|@wallrj set owner to @adiroiban|
|-|-|

I forced a trunk Python3 build and it does pass. So I guess the failure must be introduced by the changes in this branch. 

adiroban, can you investigate and post an updated patch and / or some comments here please.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

> I forced a trunk Python3 build and it does pass. So I guess the failure must be introduced by the changes in this branch.

There's a fun extra variable you overlooked.  Not every build happens in the same environment!  Buildbot pseudo-randomly selects one of two available environments (slaves) in which to run the python-3.3-tests builder.  It does this on a per-build basis.  If you look at the build page, for example <http://buildbot.twistedmatrix.com/builders/python-3.3-tests/builds/2229>, you can see which slave it selected by looking under the "BuildSlave" heading.

Differences between slave configurations can produce different results in the test suite outcome (as I'm sure you know) so be sure to check which slave is involved before you conclude that a branch is at fault.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban removed owner|
|-|-|

The error is in test_sslverify

```
twisted.trial.unittest.FailTest: not equal:
a = ("You do not have a working installation of the service_identity module: 'import of 'service_identity' halted; None in sys.modules'.  Please install it from <https://pypi.python.org/pypi/service_identity> and make sure all of its dependencies are satisfied.  Without the service_identity module and a recent enough pyOpenSSL to support it, Twisted can perform only rudimentary TLS client hostname verification.  Many valid certificate/hostname mappings may be rejected.",
 '',
 0)
b = ("You do not have a working installation of the service_identity module: 'No module named service_identity'.  Please install it from <https://pypi.python.org/pypi/service_identity> and make sure all of its dependencies are satisfied.  Without the service_identity module and a recent enough pyOpenSSL to support it, Twisted can perform only rudimentary TLS client hostname verification.  Many valid certificate/hostname mappings may be rejected.",
 '',
 0)
```

.... I run the test on my computer and they pass but I still have a lot of skipped test.... official slave only had skipped=511

I have indentity_service installed using pip (with pyopenssl and pyasn1, etc)

```
----------------------------------------------------------------------
Ran 3891 tests in 49.679s

OK (skipped=866)
(py3.3)

```

Hope that soon we can have some test-requirements.txt file which is used by buildbot and can also be used locally to create an environment similar to the one used by the buildslaves.

I have no idea what to do.

Can you please try to build again?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

> I have no idea what to do.

I tried to address this issue with my previous comment.  See also [#7355](https://github.com/twisted/twisted/issues/7355).

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_13"></a>|@wallrj set owner to @adiroiban|
|-|-|

Haha. Ok, I followed exarkun's tip of forcing two builds in quick succession and we can see that it passes on the second builder.
 * http://buildbot.twistedmatrix.com/builders/python-3.3-tests/builds/2240
 * http://buildbot.twistedmatrix.com/builders/python-3.3-tests/builds/2241

Notes:
 * All the other builds pass
 * All pyflakes warnings have gone
 * All the imports that you've replaced with namedAny look correct and make sense.

Just a few points:
 1. [branches/pyflakes-for-trial-7005/twisted/trial/test/test_tests.py](branches/pyflakes-for-trial-7005/twisted/trial/test/test_tests.py)
    1. This line looks like it can be removed
       ```
       1149             unittest.decorate(suite, unittest.TestDecorator)
       ```

       I made this change and the test still passes.

       ```
       Index: twisted/trial/test/test_tests.py
       ===================================================================
       --- twisted/trial/test/test_tests.py	(revision 42679)
       +++ twisted/trial/test/test_tests.py	(working copy)
       @@ -1144,9 +1144,7 @@
                    raise unittest.SkipTest(
                        "getrefcount not supported on this platform")
                test = self.TestCase()
       - suite = unittest.TestSuite([test])
         count1 = getrefcount(test)
       - unittest.decorate(suite, unittest.TestDecorator)
         count2 = getrefcount(test)
         self.assertEqual(count1, count2)

       ```

 2. [branches/pyflakes-for-trial-7005/twisted/trial/unittest.py](branches/pyflakes-for-trial-7005/twisted/trial/unittest.py)
    1. Adding private APIs to `__all__` seems like a bad idea. Can't you instead find the places that import those APIs from here and update their imports?

Please answer or address the numbered points above and submit another patch against log:branches/pyflakes-for-trial-7005

Thanks.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

> This line looks like it can be removed

I'm skeptical of this.  The test method docstring explains this is a test for `unittest.decorate` and line 1149 is the only line in the test that uses `unittest.decorate`.  If this line is deleted then what is the test testing?

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_15"></a>|@wallrj commented|
|-|-|

Replying to [exarkun](#note_14):
> > This line looks like it can be removed
> 
> I'm skeptical of this.  The test method docstring explains this is a test for `unittest.decorate` and line 1149 is the only line in the test that uses `unittest.decorate`.  If this line is deleted then what is the test testing?

Ah sorry, I hadn't read the docstring and I was making assumptions about how unittest.decorate worked :( 

So adiroban, please ignore comment 1 in my review above. Let me know how you feel about comment 2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_16"></a>|@adiroiban set owner to RBladesOulDam|
|-|-|

Hi

Regarding private things listed in __all__, maybe this needs a separate ticket.

Those lines are related to this comment.
```
# Grab some implementation details so tests can continue to import them from
# here, rather than being concerned with which implementation module they come
# from (is this a good idea?)
```

Rather than asking questions in comments (or leave them unanswers), I think that we should  open a ticket and answer that question.

In case having those implementation details in testsuite is a good idea, then I think that is thould be fine to keep them in __all___ ... otherwise, as suggested by Richard, we should remove those imports from testsuite and update code to directly import the private stuff.

Thanks

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_18"></a>|@wallrj set owner to @adiroiban<br>@wallrj set status to `new`|
|-|-|

Thanks again adiroban,

It looks like you answered the comments from my previous review.

I notice that the branch is missing a .misc newsfile.

I'm still not convinced that we should add all those private APIs to ``__all__``.

And I've discussed the point with exarkun and tomprince during the Boston Twisted Sprint.

We think it'd be best to drop those changes from this patch and create a followup ticket, for exposing all those private APIs from a different central module which is private to trial.

See [#7701](https://github.com/twisted/twisted/issues/7701).

If you're happy with that, I'll go ahead and add the newsfile and merge this branch without the changes to `unittest.__all__` .

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_19"></a>|@adiroiban set owner to @wallrj|
|-|-|

Thanks!

Sorry for the missing news file :)

I agree with removing the __all__ part from this patch as it should still fix other problems.

Please merge! Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_20"></a>|@wsanchez set owner to @adiroiban|
|-|-|

This has been reviewed and adiroiban has commit privs now, so removing the keyword and assigning to him to make `__all__` change and merge.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_21"></a>|@adiroiban set status to `closed`|
|-|-|

(In [43860]) Merging pyflakes-for-trial-7005: Clean pyflakes warnings for trial.

Author: adiroiban
Reviewer: wsanchez, rwall, exarkun
Fixes: [#7005](https://github.com/twisted/twisted/issues/7005)
